src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -906,6 +897,147 @@ static bool usesReceiver(BoundExpression argument)
+            if (receiver is BoundTypeExpression)
+            {
+                return receiver;
+            }
+
+            if (receiver is BoundThisReference)
+            {
+                // PROTOTYPE this was only minimally tested to observe side-effects but needs more work (in CheckValue for instance, interpolation handlers)
+                // In an extension type, `this.UnderlyingTypeMember` is replaced with `this.<UnderlyingInstanceField>$.UnderlyingTypeMember`
+                Debug.Assert(receiver.Type is not null);
+                if (receiver.Type.IsExtension && !member.ContainingType.IsExtension)
+                {
+                    var thisExtensionType = (SourceExtensionTypeSymbol)receiver.Type;
+                    Debug.Assert(thisExtensionType.UnderlyingInstanceField is not null);
+                    if (!thisExtensionType.UnderlyingInstanceField.Type.Equals(member.ContainingType, TypeCompareKind.AllIgnoreOptions))
+                    {
+                        throw ExceptionUtilities.UnexpectedValue((thisExtensionType, member.ContainingType));
+                    }
+
+                    var oldSyntax2 = _factory.Syntax;
+                    _factory.Syntax = receiver.Syntax;
+
+                    var field = _factory.Field(receiver, thisExtensionType.UnderlyingInstanceField);
+
+                    _factory.Syntax = oldSyntax2;
+                    return field;
+                }
+            }
+
+            if (member.ContainingType is not { IsExtension: true } memberExtensionType)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(receiver.Type is not null);
+            if (receiver.Type.IsExtension)
+            {
+                return receiver;
+            }
+
+            var oldSyntax = _factory.Syntax;
+            _factory.Syntax = receiver.Syntax;
+
+            tempsOpt ??= ArrayBuilder<LocalSymbol>.GetInstance();
+
+            // PROTOTYPE we'll want to allow extension member lookup and receiver adjustment for type parameter values
+            Debug.Assert(!receiver.Type.IsTypeParameter());
+            var refKind = GetRefKindForCapturedReceiverTemp(receiver, useNoneForReferenceType: true);
+            // PROTOTYPE if we decide to allow readonly members in extensions, we'll need to revisit this
+            Debug.Assert(refKind is not RefKindExtensions.StrictIn);
+            if (refKind is RefKind.RefReadOnly)
+            {
+                // Since extension members cannot be readonly, we need to force cloning before the emit layer
+                refKind = RefKind.Ref;
+            }
+
+            BoundAssignmentOperator assignmentToTemp;
+            BoundLocal temp = _factory.StoreToTemp(receiver, out assignmentToTemp, refKind);
+            tempsOpt.Add(temp.LocalSymbol);
+            Debug.Assert(temp.Type is not null);
+
+            MethodSymbol method = _factory.WellKnownMethod(WellKnownMember.System_Runtime_CompilerServices_Unsafe__As_T)
```

Comment:
>WellKnownMember.System_Runtime_CompilerServices_Unsafe__As_T [](http://example.com/codeflow?start=59&length=60)

Are we testing scenario when this helper is not available?

Reply:
Yes, `AdjustReceiver_Invocation_ObjectCreation_MissingUnsafeAs`